File: library/src/main/java/com/afollestad/aesthetic/views/AestheticTabLayout.java
package com.afollestad.aesthetic.views;

import android.content.Context;
import android.content.res.ColorStateList;
import android.graphics.Color;
import android.support.annotation.ColorInt;
import android.support.annotation.RestrictTo;
import android.support.design.widget.TabLayout;
import android.util.AttributeSet;

import com.afollestad.aesthetic.Aesthetic;
import com.afollestad.aesthetic.TabLayoutIndicatorMode;

import rx.Subscription;

import static android.support.annotation.RestrictTo.Scope.LIBRARY_GROUP;
import static com.afollestad.aesthetic.Rx.distinctToMainThread;
import static com.afollestad.aesthetic.Rx.onErrorLogAndRethrow;
import static com.afollestad.aesthetic.TintHelper.createTintedDrawable;
import static com.afollestad.aesthetic.Util.adjustAlpha;
import static com.afollestad.aesthetic.Util.isColorLight;

/** @author Aidan Follestad (afollestad) */
@RestrictTo(LIBRARY_GROUP)
public class AestheticTabLayout extends TabLayout {

  private static final float UNFOCUSED_ALPHA = 0.5f;
  private Subscription indicatorModeSubscription;
  private Subscription bgModeSubscription;
  private Subscription indicatorColorSubscription;
  private Subscription bgColorSubscription;

  public AestheticTabLayout(Context context) {
    super(context);
  }

  public AestheticTabLayout(Context context, AttributeSet attrs) {
    super(context, attrs);
  }

  public AestheticTabLayout(Context context, AttributeSet attrs, int defStyleAttr) {
    super(context, attrs, defStyleAttr);
  }

  private void setIconsColor(int color) {
    final ColorStateList sl =
        new ColorStateList(
            new int[][] {
              new int[] {-android.R.attr.state_selected}, new int[] {android.R.attr.state_selected}
            },
            new int[] {adjustAlpha(color, UNFOCUSED_ALPHA), color});
    for (int i = 0; i < getTabCount(); i++) {
      final TabLayout.Tab tab = getTabAt(i);
      if (tab != null && tab.getIcon() != null) {
        tab.setIcon(createTintedDrawable(tab.getIcon(), sl));
      }
    }
  }

  @Override
  public void setBackgroundColor(@ColorInt int color) {
    super.setBackgroundColor(color);
    int iconTextColor = isColorLight(color) ? Color.BLACK : Color.WHITE;
    setIconsColor(iconTextColor);
    setTabTextColors(adjustAlpha(iconTextColor, UNFOCUSED_ALPHA), iconTextColor);
  }

  @Override
  protected void onAttachedToWindow() {
    super.onAttachedToWindow();

    bgModeSubscription =
        Aesthetic.get()
            .tabLayoutBgMode()
            .compose(distinctToMainThread())
            .subscribe(
                mode -> {
                  if (bgColorSubscription != null) {
                    bgColorSubscription.unsubscribe();
                  }
                  switch (mode) {
                    case TabLayoutIndicatorMode.PRIMARY:
                      bgColorSubscription =
                          Aesthetic.get()
                              .primaryColor()
                              .compose(distinctToMainThread())
                              .subscribe(this::setBackgroundColor, onErrorLogAndRethrow());
                      break;
                    case TabLayoutIndicatorMode.ACCENT:
                      bgColorSubscription =
                          Aesthetic.get()
                              .accentColor()
                              .compose(distinctToMainThread())
                              .subscribe(this::setBackgroundColor, onErrorLogAndRethrow());
                      break;
                    default:
                      throw new IllegalStateException("Unimplemented bg mode: " + mode);
                  }
                },
                onErrorLogAndRethrow());

    indicatorModeSubscription =
        Aesthetic.get()
            .tabLayoutIndicatorMode()
            .compose(distinctToMainThread())
            .subscribe(
                mode -> {
                  if (indicatorColorSubscription != null) {
                    indicatorColorSubscription.unsubscribe();
                  }
                  switch (mode) {
                    case TabLayoutIndicatorMode.PRIMARY:
                      indicatorColorSubscription =
                          Aesthetic.get()
                              .primaryColor()
                              .compose(distinctToMainThread())
                              .subscribe(
                                  this::setSelectedTabIndicatorColor, onErrorLogAndRethrow());
                      break;
                    case TabLayoutIndicatorMode.ACCENT:
                      indicatorColorSubscription =
                          Aesthetic.get()
                              .accentColor()
                              .compose(distinctToMainThread())
                              .subscribe(
                                  this::setSelectedTabIndicatorColor, onErrorLogAndRethrow());
                      break;
                    default:
                      throw new IllegalStateException("Unimplemented bg mode: " + mode);
                  }
                },
                onErrorLogAndRethrow());
  }

  @Override
  protected void onDetachedFromWindow() {
    if (bgModeSubscription != null) {
      bgModeSubscription.unsubscribe();
    }
    if (indicatorModeSubscription != null) {
      indicatorModeSubscription.unsubscribe();
    }
    if (bgColorSubscription != null) {
      bgColorSubscription.unsubscribe();
    }
    if (indicatorColorSubscription != null) {
      indicatorColorSubscription.unsubscribe();
    }
    super.onDetachedFromWindow();
  }
}

File: library/src/main/java/com/afollestad/aesthetic/views/AestheticTabLayout.kt
/**
 * Designed and developed by Aidan Follestad (@afollestad)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.afollestad.aesthetic.views

import android.annotation.SuppressLint
import android.content.Context
import android.content.res.ColorStateList
import android.util.AttributeSet
import androidx.annotation.ColorInt
import com.afollestad.aesthetic.Aesthetic.Companion.get
import com.afollestad.aesthetic.ColorMode
import com.afollestad.aesthetic.utils.adjustAlpha
import com.afollestad.aesthetic.utils.distinctToMainThread
import com.afollestad.aesthetic.utils.one
import com.afollestad.aesthetic.utils.subscribeBackgroundColor
import com.afollestad.aesthetic.utils.subscribeTo
import com.afollestad.aesthetic.utils.tint
import com.afollestad.aesthetic.utils.unsubscribeOnDetach
import com.google.android.material.tabs.TabLayout

/** @author Aidan Follestad (afollestad) */
class AestheticTabLayout(
  context: Context,
  attrs: AttributeSet? = null
) : TabLayout(context, attrs) {

  companion object {
    const val UNFOCUSED_ALPHA = 0.5f
  }

  @SuppressLint("CheckResult")
  override fun setBackgroundColor(@ColorInt color: Int) {
    super.setBackgroundColor(color)
    get().toolbarIconColor()
        .one()
        .subscribeTo(::setIconsColor)
        .unsubscribeOnDetach(this)
    get().toolbarTitleColor()
        .one()
        .subscribeTo { setTabTextColors(it.adjustAlpha(UNFOCUSED_ALPHA), it) }
        .unsubscribeOnDetach(this)
  }

  override fun onAttachedToWindow() {
    super.onAttachedToWindow()

    get().toolbarIconColor()
        .subscribeTo(::setIconsColor)
        .unsubscribeOnDetach(this)

    get().toolbarTitleColor()
        .subscribeTo { setTabTextColors(it.adjustAlpha(UNFOCUSED_ALPHA), it) }
        .unsubscribeOnDetach(this)

    get().tabLayoutBackgroundMode()
        .distinctToMainThread()
        .flatMap {
          when (it) {
            ColorMode.PRIMARY -> get().colorPrimary()
            ColorMode.ACCENT -> get().colorAccent()
          }
        }
        .distinctToMainThread()
        .subscribeBackgroundColor(this@AestheticTabLayout)
        .unsubscribeOnDetach(this)

    get().tabLayoutIndicatorMode()
        .distinctToMainThread()
        .flatMap {
          when (it) {
            ColorMode.PRIMARY -> get().colorPrimary()
            ColorMode.ACCENT -> get().colorAccent()
          }
        }
        .distinctToMainThread()
        .subscribeTo(::setSelectedTabIndicatorColor)
        .unsubscribeOnDetach(this)
  }

  private fun setIconsColor(color: Int) {
    val sl = ColorStateList(
        arrayOf(
            intArrayOf(-android.R.attr.state_selected),
            intArrayOf(android.R.attr.state_selected)
        ),
        intArrayOf(
            color.adjustAlpha(UNFOCUSED_ALPHA),
            color
        )
    )
    for (i in 0 until tabCount) {
      val tab = getTabAt(i)
      if (tab != null && tab.icon != null) {
        tab.icon = tab.icon.tint(sl)
      }
    }
  }
}
